crypto/src/proofs/transparent.rs

```diff
@@ -56,63 +57,46 @@ impl SpendProof {
             return Ok(());
         }
 
-        // Note commitment integrity.
-        let s_component_transmission_key = self.note.transmission_key_s();
-        let note_commitment_test = note::commitment(
+        gadgets::note_commitment_integrity(
             self.note.note_blinding(),
             self.note.value(),
             self.note.diversified_generator(),
-            s_component_transmission_key,
-            self.note.clue_key(),
-        );
-
-        if self.note_commitment_proof.commitment() != note_commitment_test {
-            return Err(anyhow!("note commitment mismatch"));
-        }
+            self.note.transmission_key_s(),
+            self.note.clue_key().clone(),
+            self.note_commitment_proof.commitment(),
+        )?;
 
         // Merkle path integrity.
         self.note_commitment_proof
             .verify(anchor)
             .map_err(|_| anyhow!("merkle root mismatch"))?;
 
-        // Value commitment integrity.
-        if self.note.value().commit(self.v_blinding) != balance_commitment {
-            return Err(anyhow!("value commitment mismatch"));
-        }
+        gadgets::value_commitment_integrity(
+            balance_commitment,
+            self.v_blinding,
+            self.note.value(),
+        )?;
 
-        // The use of decaf means that we do not need to check that the
-        // diversified basepoint is of small order. However we instead
-        // check it is not identity.
-        if self.note.diversified_generator().is_identity() || self.ak.is_identity() {
+        gadgets::diversified_basepoint_not_identity(self.note.diversified_generator().clone())?;
+        if self.ak.is_identity() {
             return Err(anyhow!("unexpected identity"));
         }
 
-        // Nullifier integrity.
-        if nullifier
-            != self.nk.derive_nullifier(
-                self.note_commitment_proof.position(),
-                &self.note_commitment_proof.commitment(),
-            )
-        {
-            return Err(anyhow!("bad nullifier"));
-        }
+        gadgets::nullifier_integrity(
+            nullifier,
+            self.nk,
+            self.note_commitment_proof.position(),
+            self.note_commitment_proof.commitment(),
+        )?;
 
-        // Spend authority.
-        let rk_bytes: [u8; 32] = rk.into();
-        let rk_test = self.ak.randomize(&self.spend_auth_randomizer);
-        let rk_test_bytes: [u8; 32] = rk_test.into();
-        if rk_bytes != rk_test_bytes {
-            return Err(anyhow!("invalid spend auth randomizer"));
-        }
+        gadgets::check_spend_authority(self.spend_auth_randomizer, rk, self.ak)?;
```

Comment:
Naming nitpick: this isn't really checking that the prover has spend authority, right? The actual spend authority is checked by the auth sig, and this is proving that the provided randomized verification key is a correctly formed randomization of the right authorizing key.

Could we find a different name for this gadget that would capture this idea?

Not sure what the best name is, but what about something like `rk_integrity` or something more descriptive?